core/src/serialize_ext.rs

```diff
@@ -0,0 +1,80 @@
+use cid::Cid;
+use serde::Serialize;
+
+/// SerializeExt is a trait for serialization, deserialization, CID calculation assuming dag-cbor.
+pub trait SerializeExt: Serialize {
+    /// Error type for SerializeExt
+    type Error;
+
+    /// Serialize to dag-json
+    fn to_json(&self) -> Result<String, Self::Error>;
+
+    /// Serialize to dag-cbor
+    fn to_cbor(&self) -> Result<Vec<u8>, Self::Error>;
+
+    /// CID of Serialize to dag-cbor
+    fn to_cid(&self) -> Result<Cid, Self::Error>;
+
+    /// Serialize to dag-cbor and return the CID and bytes
+    fn to_dag_cbor_block(&self) -> Result<(Cid, Vec<u8>), Self::Error>;
+}
+
+impl<T: Serialize> SerializeExt for T {
+    type Error = anyhow::Error;
```

Comment:
Why are we using anyhow here?

Reply:
It looks like the ipld crates use two different error types and don't just use a serde error, so we need to provide something that matches both. Should we use an error that passes through the variants or is using anyhow okay?

```rust
// ipld_dag_json
#[derive(Debug)]
pub enum EncodeError {
    Message(String),
}

// ipld_dag_cbor
#[derive(Debug)]
pub enum EncodeError<E> {
    /// Custom error message.
    Msg(String),
    /// IO Error.
    Write(E),
}
```

Reply:
Lets use [thiserror](https://docs.rs/thiserror/latest/thiserror/) and create a variant for each type of error that can be produced. Thiserror make this very easy. With this approach we expose the underlying errors which is what we want in the context of a library trait.